flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -929,6 +929,10 @@ private void restoreAndInvoke(TaskInvokable finalInvokable) throws Exception {
                     new TaskExecutionState(executionId, ExecutionState.RUNNING));
 
             runWithSystemExitMonitoring(finalInvokable::invoke);
+
+            taskManagerActions.notifyTaskFinishing(executionId);
```

Comment:
Would it be simpler to pass `taskManagerActions` to the constructor of `StreamTask` via `RuntimeEnvironment` and let `StreamTask` invoke `notifyTaskFinishing` directly after receiving `END_OF_DATA`?

Reply:
Yes, it would be simpler and I've updated the PR based on your advice. Could you have another review?